extensions/oidc-devservices-keycloak/deployment/src/main/java/io/quarkus/devservices/keycloak/deployment/KeycloakDevServicesProcessor.java

```diff
@@ -0,0 +1,395 @@
+        realm.getClients().add(createClient(oidcClientId, oidcClientSecret));
+        for (Map.Entry<String, String> entry : users.entrySet()) {
+            realm.getUsers().add(createUser(entry.getKey(), entry.getValue(), getUserRoles(entry.getKey())));
+        }
+
+        WebClient client = KeycloakDevServicesUtils.createWebClient();
+        try {
+            String token = KeycloakDevServicesUtils.getPasswordAccessToken(client,
+                    keycloakUrl + "/realms/master/protocol/openid-connect/token",
+                    "admin-cli", null, "admin", "admin");
+
+            client.postAbs(keycloakUrl + "/admin/realms")
+                    .putHeader(HttpHeaders.CONTENT_TYPE.toString(), "application/json")
+                    .putHeader(HttpHeaders.AUTHORIZATION.toString(), "Bearer " + token)
+                    .sendBuffer(Buffer.buffer().appendString(JsonSerialization.writeValueAsString(realm)))
+                    .await().indefinitely();
+        } catch (Throwable t) {
+            LOG.errorf("Realm %s can not be created: %s", realm.getRealm(), t.getMessage());
+        } finally {
+            client.close();
+        }
+    }
+
+    private Map<String, String> getUsers(Map<String, String> configuredUsers) {
+        if (configuredUsers.isEmpty()) {
+            Map<String, String> users = new LinkedHashMap<String, String>();
+            users.put("alice", "alice");
+            users.put("bob", "bob");
+            return users;
+        } else {
+            return configuredUsers;
+        }
+    }
+
+    private String[] getUserRoles(String user) {
+        String roles = config.devservices.roles.get(user);
+        return roles == null ? ("alice".equals(user) ? new String[] { "admin", "user" } : new String[] { "user" })
+                : roles.split(",");
+    }
+
+    private RealmRepresentation createRealmRep() {
+        RealmRepresentation realm = new RealmRepresentation();
+
+        realm.setRealm(config.devservices.realmName);
+        realm.setEnabled(true);
+        realm.setUsers(new ArrayList<>());
+        realm.setClients(new ArrayList<>());
+
+        RolesRepresentation roles = new RolesRepresentation();
+        List<RoleRepresentation> realmRoles = new ArrayList<>();
+
+        roles.setRealm(realmRoles);
+        realm.setRoles(roles);
+
+        if (config.devservices.roles.isEmpty()) {
+            realm.getRoles().getRealm().add(new RoleRepresentation("user", null, false));
+            realm.getRoles().getRealm().add(new RoleRepresentation("admin", null, false));
+        } else {
+            for (String role : config.devservices.roles.values()) {
+                realm.getRoles().getRealm().add(new RoleRepresentation(role, null, false));
+            }
+        }
+        return realm;
+    }
+
+    private ClientRepresentation createClient(String clientId, String oidcClientSecret) {
+        ClientRepresentation client = new ClientRepresentation();
+
+        client.setClientId(clientId);
```

Comment:
Should this configure CORS to automatically allow requests from localhost:8080 (or whatever the HTTP/HTTPS listeners are bound to)?

I tried to enable OIDC in Swagger UI and I could not figure this bit out, as keycloak would not send back an Access-Control-Allowed-Origins header.

Reply:
@stuartwdouglas Hi Stuart - do you mean you could not authenticate to Keycloak via Keycloak's authorization endpoint ? That endpoint does not support CORS - I'd propose to look at wiring in Swagger UI in a follow up task, such as #18597

Reply:
Why doesn't it support cors? I thought keycloak.js would send cross origin requests to this endpoint to perform authentication?